content/articles/mocking-with-junit-and-mockito-the-why-and-how/index.md

```diff
@@ -0,0 +1,1011 @@
+```studentRepository``` whose ```matricNo``` matches the one specified. If we find a match, we obtain the student record from the ```studentRepository```; else, we throw an exception. If the ```gender``` field in the student record retrieved from the database is ```MALE```, a call is made to the ```returnAvailableMaleSpace``` method in the ```HostelRepository``` class. The implementation of ```returnAvailableMaleSpace``` is seen in the second snippet. We simply return the first available bed-space in ```availableBedSpacesForMales``` (represented as a queue) using the ```poll``` method. We throw an exception if there is no available bed space (as shown in the catch clause). At the end of the ```returnAvailableBedSpacesForMales``` method, control is returned back to the ```assignBedSpace``` method.
+Next, we set the ```bedSpaceId``` field in the ```student``` object with the value of the ```id``` field in the returned ```bedSpace``` object. Finally, the method concludes by setting the ```isEmpty``` field in the returned ```bedSpace```
+object to false.
+
+>The implementation of ```returnAvailableFemaleSpace``` method is similar to the implementation of ```returnAvailableMaleSpace```. The only difference is that ```returnAvailableFemaleSpace``` retrives the first available bedspace from the ```availableBedSpacesForMales``` queue.
+
+
+``` java
+@Override
+public void assignBedSpace(StudentDto studentDto) throws Exception {
+    Student student = studentRepository.findById(studentDto.getId()).orElseThrow(()->
+            new HostelManagementException("student with specified matric number not found!"));
+    BedSpace bedSpace;
+    if(studentDto.getGender()== Gender.FEMALE){
+        bedSpace = hostelRepository.returnAvailableFemaleSpace();
+    }
+    else{
+        bedSpace = hostelRepository.returnAvailableMaleSpace();
+    }
+    student.setBedSpaceId(bedSpace.getId());
+    bedSpace.setEmpty(false);
+}
+```
+
+
+```java
+public BedSpace returnAvailableMaleSpace() throws NoAvailableBedspaceException {
+    try {
+      return availableBedSpacesForMales.poll();
+    }
+    catch(NoSuchElementException exception){
+      throw new NoAvailableBedspaceException(exception.getMessage());
+    }
+}
+```
+
+
+Our test class for verifying the number of invocation of specific methods of mocked objects is shown below:
+
+```java
+public class StudentServiceImpl_VerifyTests {
+    private StudentService studentService;
+    @Spy
+    private StudentRepository studentRepository;
+    @Spy
+    private HostelRepository hostelRepository;
+    private LocalDateTime time;
+    private Student student;
+
+    @BeforeEach
+    void setUp() {
+        time = null;
+        studentService = new StudentServiceImpl(studentRepository, hostelRepository);
+    }
+}
+```
+
+First, we create spies of ```studentRepository```and ```hostelRepository```. Next, we inject them into an object of
+```studentServiceImpl```
+
+Next, our test case is shown below:
```

Comment:
Remove the word "Next" from this statement.

Reply:
Done